llama-index-core/llama_index/core/agent/introspective/step.py

```diff
@@ -0,0 +1,312 @@
+        infer defaults for a variety of parameters, including the LLM,
+        if they are not specified.
+
+        """
+        tools = tools or []
+
+        llm = llm or Settings.llm
+        if callback_manager is not None:
+            llm.callback_manager = callback_manager
+
+        if system_prompt is not None:
+            if prefix_messages is not None:
+                raise ValueError(
+                    "Cannot specify both system_prompt and prefix_messages"
+                )
+            prefix_messages = [ChatMessage(content=system_prompt, role="system")]
+
+        prefix_messages = prefix_messages or []
+
+        return cls(
+            tools=tools,
+            tool_retriever=tool_retriever,
+            main_agent_worker=main_agent_worker,
+            reflective_agent_worker=reflective_agent_worker,
+            llm=llm,
+            prefix_messages=prefix_messages,
+            verbose=verbose,
+            max_function_calls=max_function_calls,
+            callback_manager=callback_manager,
+            **kwargs,
+        )
+
+    def initialize_step(self, task: Task, **kwargs: Any) -> TaskStep:
+        """Initialize step from task."""
+        # temporary memory for new messages
+        main_memory = ChatMemoryBuffer.from_defaults()
+        reflective_memory = ChatMemoryBuffer.from_defaults()
+
+        # initialize task state
+        task_state = {
+            "main": {
+                "memory": main_memory,
+                "sources": [],
+            },
+            "reflection": {"memory": reflective_memory, "sources": []},
+        }
+        task.extra_state.update(task_state)
+
+        return TaskStep(
+            task_id=task.task_id,
+            step_id=str(uuid.uuid4()),
+            input=task.input,
+        )
+
+    def get_tools(self, input: str) -> List[AsyncBaseTool]:
+        """Get tools."""
+        return [adapt_to_async_tool(t) for t in self._get_tools(input)]
+
+    def get_all_messages(self, task: Task) -> List[ChatMessage]:
+        return (
+            self.prefix_messages
+            + task.memory.get()
+            + task.extra_state["main"]["memory"].get_all()
+            + task.extra_state["reflection"]["memory"].get_all()
+        )
+
+    @trace_method("run_step")
+    def run_step(self, step: TaskStep, task: Task, **kwargs: Any) -> TaskStepOutput:
+        """Run step."""
+        # run main agent
+        main_agent = self._main_agent_worker.as_agent()
+        main_agent_response = main_agent.chat(
+            task.input
+        )  # or should i use step.input here?
+        task.extra_state["main"]["sources"] = main_agent_response.sources
+        task.extra_state["main"]["memory"] = main_agent.memory
+        print(f"MAIN AGENT MEMORY: {main_agent.memory}", flush=True)
+
+        # run reflective agent
+        reflective_agent = self._reflective_agent_worker.as_agent()
+        reflective_agent_response = reflective_agent.chat(main_agent_response.response)
```

Comment:
the only thing the reflective agent has for context here is the response from the first agent. Shouldn't it also have some context on what the original query was too?

Reply:
Yeah, that's a good point!

Reply:
This is the only one that I haven't resolved yet. I do pass the chat history to the ToolInteractiveReflectionAgent, but not to the critique agent which gets the task of reflection with tools.

To do this would require a bit of re-jigging on the prompts for the CritiqueAgent and then ofc the passing of the chat history, which can be done as prompt str.

I think it would be okay to do this in a future release of this package, in order to get this package out sooner. Thoughts?